### ISSUE TITLE
assertion in solveWellEq for flow.

### ISSUE THREAD

**Issue by dr-robertk** [OPM/opm-autodiff#472]
When running flow in parallel an assertion is thrown in `BlackoilModelBase_impl.hpp` line 848 where a method `solveWellEq` is called. 

The assertion is:
```
 OPM/master/opm-autodiff/eigen3-installed/include/eigen3/Eigen/src/Core/DenseCoeffsBase.h:394: Eigen::DenseCoeffsBase<Derived, 1>::Scalar& Eigen::DenseCoeffsBase<Derived, 1>::operator()(Eigen::DenseCoeffsBase<Derived, 1>::Index) [with Derived = Eigen::Matrix<int, -1, 1>; Eigen::DenseCoeffsBase<Derived, 1>::Scalar = int; Eigen::DenseCoeffsBase<Derived, 1>::Index = long int]: Assertion `index >= 0 && index < size()' failed.
```

**Comment by dr-robertk** [comment]
This assertion is also thrown with flow, i.e. in serial runs.

**Comment by blattms** [comment]
Please tell me the options your are using for your run.

**Comment by dr-robertk** [comment]
It doesn't seem to bother anybody at least I can't debug flow right now, because when run without NDEBUG I get an assertion somewhere in solveWellEq. 

flow: opm-autodiff/eigen3-installed/include/eigen3/Eigen/src/Core/DenseCoeffsBase.h:394: Eigen::DenseCoeffsBase<Derived, 1>::Scalar& Eigen::DenseCoeffsBase<Derived, 1>::operator()(Eigen::DenseCoeffsBase<Derived, 1>::Index) [with Derived = Eigen::Matrix<int, -1, 1>; Eigen::DenseCoeffsBase<Derived, 1>::Scalar = int; Eigen::DenseCoeffsBase<Derived, 1>::Index = long int]: Assertion `index >= 0 && index < size()' failed.
Aborted

**Comment by blattms** [comment]
It bothers me, but I have to repeat myself:
> Please tell me the options your are using for your run and the model.

**Comment by dr-robertk** [comment]
dune 2.4 with opm latest and greatest: 

I run 

flow SPE9.DATA. 

gcc 4.8.4 with -g -Wall -DDUNE_DEVEL_MODE -Wfatal-errors -pipe -Wno-unknown-pragmas

**Comment by bska** [comment]
> when run without `NDEBUG` I get an assertion somewhere in `solveWellEq` for `flow` on SPE9.

I can't reproduce that symptom locally so you'll need to provide more details.  For the record, I just completed a `flow` run using
```sh
${exe} path/to/SPE9.DATA
```
with `${exe}` being the full path to my debug version of `flow` built with GCC 4.8.4, [Eigen 3.2.1](https://github.com/OPM/eigen3/commit/4454cb96dad2c57d1574899a759e1062c13a13b0), sequential configuration, full libstdc++ debug mode, targeting Dune 2.3.1 on opm-data's SPE9 as of OPM/opm-data@43a0c02.  The last few log lines are as follows
```
---------------    Simulation step number 89    ---------------
      Current time (days)     890
      Current stepsize (days) 10
      Total time (days)       900

Substep( 0 ), try with stepsize 10 (days).

Iter W-FLUX(W)  W-FLUX(O)  W-FLUX(G)
   0  1.196e-07  1.694e-06  1.739e-05
well converged iter: 0

Iter  MB(WATER)   MB(OIL)    MB(GAS)       CNVW       CNVO       CNVG   W-FLUX(W)  W-FLUX(O)  W-FLUX(G)
   0  1.340e-05  2.947e-04  2.355e-03  1.976e-03  3.161e-03  8.617e-03  1.242e-07  1.744e-06  1.759e-05
   1  3.155e-07  3.783e-08  5.491e-06  1.002e-04  2.681e-04  4.510e-03  2.930e-08  2.475e-07  4.999e-06
Overall linear iterations used: 6
Substep( 0 ) finished at time 900 (days).

Sub steps started at time = 890 (days)
 step[ 0 ] = 10 (days)
sub steps end time = 900 (days)
Suggested next step size = 30 (days)
Fully implicit solver took: 4.63985 seconds.


================    End of simulation     ===============

Total time taken (seconds):  710.402
Solver time (seconds):       709.111
Overall Newton Iterations:   196
Overall Linear Iterations:   870
```

In other words, whatever the problem is, it does not present in my setup.  Maybe you can get **some** useful information--at least a stack trace--if you set
```gdb
catch throw
```
before you `run` the simulation case in GDB.

**Comment by blattms** [comment]
Might this depend on the Eigen version? I recall that @dr-robertk is using an older version.

**Comment by blattms** [comment]
I recall getting something similar with one of my preliminary parallel version. I think it had something to do with wells that do not have any perforations. Can that happen in flow? Then there might be some serious issues because of missing checks `if (wells.conn_pos[w]==wells.conn_pos[w-1])` in our code.

**Comment by blattms** [comment]
BTW I assume the Eigen matrix passed to SparseLU has either zero or negative size in this case.

**Comment by bska** [comment]
> Might this depend on the Eigen version?

That's certainly possible.

> I recall that @dr-robertk is using an older version.

Right.  He **did** say earlier that he was

>> using the dune 2.4 and the auto installed Eigen.

which, according to [`CMakeLists.txt`](https://github.com/OPM/opm-autodiff/blob/master/CMakeLists.txt#L111) would mean commit OPM/eigen3@9e788db.  At least from that point of view his Eigen installation should match my own.

> I recall getting something similar with one of my preliminary parallel versions. I think it had something to do with wells that do not have any perforations. Can that happen in flow?

Yes, although it should not happen on SPE9.

**Comment by blattms** [comment]
Indeed the matrix is null:
```gdb
(gdb) 
1473	                Jn[0].toSparse(Jn0);
(gdb) p Jn0.size()
Cannot evaluate function -- may be inlined
(gdb) p Jn0
$1 = {<Eigen::SparseMatrixBase<Eigen::SparseMatrix<double, 0, int> >> = {<Eigen::EigenBase<Eigen::SparseMatrix<double, 0, int> >> = {<No data fields>}, 
    m_isRValue = false}, m_outerSize = 0, m_innerSize = 0, 
  m_outerIndex = 0x2abd3e0, m_innerNonZeros = 0x0, m_data = {m_values = 0x0, 
    m_indices = 0x0, m_size = 0, m_allocatedSize = 0}}
(gdb) p Jn0.rows()
$2 = 0
(gdb) p Jn0.cols()
$3 = 0
(gdb) 
```

**Comment by blattms** [comment]
My previous answer is rubbish as Jn0 is not initialized, yet. One line below the above one, it has 104 rows and 104 columns.

**Comment by blattms** [comment]
I cannot repoduce it either. I am using Eigen 3.2.2.

**Comment by dr-robertk** [comment]
This problem has vanished. (That's the kind of problem I like...)
